src/levenberg.jl

```diff
@@ -0,0 +1,424 @@
+end
+
+function SciMLBase.__init(prob::NonlinearProblem{uType, iip}, alg::LevenbergMarquardt,
+                          args...;
+                          alias_u0 = false,
+                          maxiters = 1000,
+                          abstol = 1e-6,
+                          internalnorm = DEFAULT_NORM,
+                          kwargs...) where {uType, iip}
+    if alias_u0
+        u = prob.u0
+    else
+        u = deepcopy(prob.u0)
+    end
+    f = prob.f
+    p = prob.p
+    if iip
+        fu = zero(u)
+        f(fu, u, p)
+    else
+        fu = f(u, p)
+    end
+    uf, linsolve, J, du_tmp, jac_config = jacobian_caches(alg, f, u, p, Val(iip))
+
+    if u isa Number
+        DᵀD = alg.min_damping_D
+    else
+        d = similar(u)
+        d .= alg.min_damping_D
+        DᵀD = Diagonal(d)
+    end
+
+    loss = internalnorm(fu)
+    JᵀJ = zero(J)
+    v = zero(u)
+    a = zero(u)
+    tmp_vec = zero(u)
+    v_old = zero(u)
+    δ = zero(u)
+    make_new_J = true
+    fu_tmp = zero(fu)
+    mat_tmp = zero(J)
+
+    return LevenbergMarquardtCache{iip}(f, alg, u, fu, p, uf, linsolve, J, du_tmp,
+                                        jac_config,
+                                        1, false, maxiters, internalnorm,
+                                        ReturnCode.Default, abstol, prob, DᵀD, JᵀJ,
+                                        alg.damping_initial, alg.damping_increase_factor,
+                                        v, a, tmp_vec, v_old, loss, δ, loss, make_new_J,
+                                        fu_tmp, mat_tmp)
+end
+function perform_step!(cache::LevenbergMarquardtCache{true})
+    @unpack fu, f, make_new_J = cache
+    if iszero(fu)
+        cache.force_stop = true
+        return nothing
+    end
+    if make_new_J
+        jacobian!(cache.J, cache)
+        mul!(cache.JᵀJ, cache.J', cache.J)
+        cache.DᵀD .= max.(cache.DᵀD, Diagonal(cache.JᵀJ))
+        cache.make_new_J = false
+    end
+    @unpack u, p, λ, JᵀJ, DᵀD, J, alg, linsolve = cache
+
+    # Usual Levenberg-Marquardt step ("velocity").
+    # The following lines do: cache.v = -cache.mat_tmp \ cache.fu_tmp
+    mul!(cache.fu_tmp, J', fu)
+    @. cache.mat_tmp = JᵀJ + λ * DᵀD
+    linres = dolinsolve(alg.precs, linsolve, A = cache.mat_tmp, b = _vec(cache.fu_tmp),
+                        linu = _vec(cache.du_tmp), p = p, reltol = cache.abstol)
+    cache.linsolve = linres.cache
+    @. cache.v = -cache.du_tmp
+
+    # Geodesic acceleration (step_size = v + a / 2).
+    @unpack v, alg = cache
+    h = alg.finite_diff_step_geodesic
+    f(cache.fu_tmp, u .+ h .* v, p)
+
+    # The following lines do: cache.a = -J \ cache.fu_tmp
+    mul!(cache.du_tmp, J, v)
+    @. cache.fu_tmp = (2 / h) * ((cache.fu_tmp - fu) / h - cache.du_tmp)
+    linres = dolinsolve(alg.precs, linsolve, A = J, b = _vec(cache.fu_tmp),
+                        linu = _vec(cache.du_tmp), p = p, reltol = cache.abstol)
```

Comment:
We may want to split the preconditioners between these two. But that's for follow up.